src/FromYAML.hs

```diff
@@ -143,69 +221,65 @@ instance p ~ Nickname => FromJSON (J (Message p)) where
             _ -> []
         [(a, v)]
 
-instance FromJSON (J Thumbnail) where
-  parseJSON =
-    withObject "Thumbnail" $ \j -> do
-      J thumbnailSide <- j .: "side"
-      J thumbnailPic <- j .: "pic"
-      (getJm -> thumbnailLink) <- j .:? "link"
-      thumbnailClass <- fmap (fromMaybe mempty) $ optional $
-        fmap getJset (j .: "class") <|>
-        fmap (Set.singleton . getJ) (j .: "class")
-      (getJm -> thumbnailCaption) <- j .:? "caption"
-      return $ J Thumbnail{..}
+----------------------------------------------------------------------------
+-- Post instances
+----------------------------------------------------------------------------
 
-instance FromJSON (J Side) where
+instance (id ~ (), p ~ Nickname) => FromJSON JPost (Post id p) where
   parseJSON =
-    withText "Side" $ \t ->
-      J <$> case t of
-        "left" -> return SideLeft
-        "right" -> return SideRight
-        _ -> fail "Unknown side"
+    withObject "Post" $ \j -> do
+      let postId = ()
+      postTitle <- j .: "title"
+      postDates <- j .: "date"
+      postAuthor <- j .: "author"
+      postBody <- j .: "body"
+      return $ Post{..}
 
-instance FromJSON (J Class) where
-  parseJSON =
-    withText "Class" $ return . J . Class
+instance FromJSON JPost PostDates where
+  parseJSON = \j -> singleDate j
+    where
+      singleDate j = do
+        date <- parseJSON j
+        return $ PostDates date
 
-instance FromJSON (J Caption) where
+instance FromJSON JPost PostBody where
   parseJSON =
-    withText "Caption" $ pure . J . Caption . parseMMark
+    withArray "PostBody" $ \contentParts -> do
+      contentParts' <-
+        case nonEmpty (Foldable.toList contentParts) of
+          Nothing -> fail "empty content parts"
+          Just a -> return a
+      PostBody <$> traverse parseJSON contentParts'
+
+----------------------------------------------------------------------------
+-- People instances
+----------------------------------------------------------------------------
 
-instance FromJSON (J Person) where
+instance FromJSON JPeople Person where
   parseJSON =
     withObject "Person" $ \j -> do
-      (getJs -> pNicks) <- j .: "nicks"
-      J pName <- j .: "name"
-      J pAlias <- j .: "alias"
-      (getJm -> pPic) <- j .:? "pic"
-      (getJm -> pLink) <- j .:? "link"
-      J pRole <- j .: "role"
-      return $ J Person{..}
+      pNicks <- j .: "nicks"
+      pName <- j .: "name"
+      pAlias <- j .: "alias"
+      pPic <- j .:? "pic"
+      pLink <- j .:? "link"
+      pRole <- j .: "role"
+      return $ Person{..}
 
-instance FromJSON (J Nickname) where
-  parseJSON =
-    withText "Nickname" $ return . J . Nickname
+----------------------------------------------------------------------------
+-- Standard types
+----------------------------------------------------------------------------
 
-instance FromJSON (J Name) where
-  parseJSON =
-    withText "Name" $ return . J . Name
+-- TODO: all of these are bad
```

Comment:
Why?

Reply:
They are ugly, the combination of type applications had to be guessed by trial and error, `fmap Set.fromList` is inefficient, `parseJSON = parseJSONList` only works for lists, and overall there should be a way to define them completely or almost completely uniformly.